jerry-core/api/jerry.c

```diff
@@ -207,22 +197,33 @@ jerry_cleanup (void)
 
   jmem_finalize ();
   jerry_make_api_unavailable ();
-
-  if (JERRY_CONTEXT (user_context_deinit_cb))
-  {
-    JERRY_CONTEXT (user_context_deinit_cb) (JERRY_CONTEXT (user_context_p));
-  }
 } /* jerry_cleanup */
 
 /**
  * Retrieve user context.
  *
- * @return the user-provided context-specific pointer
+ * @param manager_p pointer to the manager whose user context item should be returned.
+ *
+ * @return the user-provided context-specific pointer for the given manager, creating such a pointer if none was found.
  */
-void *
-jerry_get_user_context (void)
+jerry_user_context_item_t *
+jerry_get_user_context (const jerry_user_context_manager_t *manager_p)
 {
-  return JERRY_CONTEXT (user_context_p);
+  jerry_user_context_item_t *item_p;
+
+  for (item_p = JERRY_CONTEXT (user_context_p); item_p; item_p = item_p->next_p)
+  {
+    if (item_p->manager_p == manager_p)
+    {
+      return item_p;
+    }
+  }
+
```

Comment:
Hmm, this way `init_cb` is only called after the first `jerry_get_user_context()` call? That's a fundamental difference from the approach from #1802, where all `init_cb`s were called during `jerry_init`.

I like the approach of #1802 a little bit better, because it enables components built on top of it to automagically (but optionally) initialize together with JerryScript itself – just add the sources of the component to your project's build and you're done. This is useful for example if the component adds APIs to the global object. If we go with the approach in this PR, you'd manually have to call a `my_component_api_add_to_global_object()` after `jerry_init()`.